### PR TITLE
support configure hystrix thread isolation in backend service level for zuul

### PR DESCRIPTION
**Pull request by rodneyqin**
modify AbstractRibbonCommand to support configure hystrix thread isolation in backend service level

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@rodneyqin Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=2725)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-netflix&pullRequestId=2725) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by rodneyqin** [comment]
@ryanjbaxter How should I show the test? Can I upload the screen shot of turbine which shows different hystrix statistics base on each backend service on zull gateway?

File: spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/support/AbstractRibbonCommand.java [hunk 1]
```diff
@@ -123,9 +123,33 @@ protected static Setter getSetter(final String commandKey, ZuulProperties zuulPr
 	protected static Setter getSetter(final String commandKey,
 			ZuulProperties zuulProperties, IClientConfig config) {
 
-		// @formatter:off
-		Setter commandSetter = Setter.withGroupKey(HystrixCommandGroupKey.Factory.asKey("RibbonCommand"))
-								.andCommandKey(HystrixCommandKey.Factory.asKey(commandKey));
+		RequestContext requestContext = RequestContext.getCurrentContext();
+		String routeId = (String)requestContext.get("proxy");
```
Comment (spencergibb) [inline spring-cloud-netflix-zuul/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/support/AbstractRibbonCommand.java]: There shouldn't be a need for this. The service is available in `IClientConfig config`

**Comment by ryanjbaxter** [comment]
Can you extend this class and make the sure the setter is correct?

**Comment by yongsungyoon** [comment]
@ryanjbaxter, @spencergibb 
This PR is not needed. Suggested feature already exists in master branch and 1.4.x branch.
`AbstractRibbonCommand.java`
```
} else if (zuulProperties.getThreadPool().isUseSeparateThreadPools()) {
    final String threadPoolKey = zuulProperties.getThreadPool().getThreadPoolKeyPrefix() + commandKey; 			 
    commandSetter.andThreadPoolKey(HystrixThreadPoolKey.Factory.asKey(threadPoolKey));
}	
```
As you can see above, if zuul is running on hystrix thread isolation, Zuul is setting commandKey as threadPoolKey just with prefix (default prefix is `zuul-`). And commandGroupKey is only used as threadpoolKey if threadPoolKey is not specified.

As a result, suggested feature by this PR is already provided. Use can set hystrix properties as service id level like below.
```
hystrix.command.zuul-<service id>.execution.isolation.thread.timeoutInMilliseconds
```
This feature is already documented [here](http://cloud.spring.io/spring-cloud-netflix/1.4.x/single/spring-cloud-netflix.html#how-to-configure-hystrix-thread-pools)

@rodneyqin, could you check the document above ?

**Comment by YunGeol** [comment]
@yongsungyoon  Fantastic! 
Everything is good, but one thing to point out is ...
```
zuul.threadPoolKeyPrefix=zuulgw    # It is not correct.
zuul.threadPool.threadPoolKeyPrefix: zuulgw # It is correct.
```
([document](http://cloud.spring.io/spring-cloud-netflix/1.4.x/single/spring-cloud-netflix.html#how-to-configure-hystrix-thread-pools))
I appreciate your efforts, yongsungyoon.
I found what I wanted because of you.

**Comment by yongsungyoon** [comment]
@YunGeol thanks. I've updated my comments above

**Pull request closed without merging**